icefoganalytics/travel-authorization
# Unable to Add Agent

Fixes https://github.com/ytgov/travel-authorization/issues/105

Relates to:

- https://governmentofyukon.slack.com/archives/C04ETKKJS1X/p1726786971993799

# Context

**Describe the bug** I am unable to add a new agency to the application. The button is greyed out. I have all the system roles

**To Reproduce** Steps to reproduce the behavior:

1. Go to 'Administration'
2. Click on 'Travel Agents' in the look-up tables
3. note the add new agent button is greyed out on the top left
4. See error

**Expected behavior** As travel desk user and system administrator I should have the ability to add or edit an agency

**Screenshots** ![image](https://private-user-images.githubusercontent.com/89539008/351526393-7f911363-d3c3-4ef0-8691-3b8d10c56778.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjY4NTYxMzUsIm5iZiI6MTcyNjg1NTgzNSwicGF0aCI6Ii84OTUzOTAwOC8zNTE1MjYzOTMtN2Y5MTEzNjMtZDNjMy00ZWYwLTg2OTEtM2I4ZDEwYzU2Nzc4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTIwVDE4MTAzNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTY4MTQzMzcyMGM5ZjUzODcwMjA1MDBkNDcwZDFlMGIxZGJlOWY3MmRjZTJmZGJhYmUzZGE2ODUxNTQzOGNjYzgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.8Z2Enm0sTSa287kwzDlXCUVnRa7R4rrC6mZgrg-yf0o)

See https://governmentofyukon.slack.com/archives/C04ETKKJS1X/p1726786971993799
TravelDeskTravelAgent model is renamed to TravelDeskTravelAgency

Ensure model has these fields:
- agency name - non null
- contact name
- contact email
- contact phone number
- extra info

# Implementation

1. Rename "travelDeskTravelAgent" table to "travel_desk_travel_agencies".
2. Migrate all travel agency api class to newly created standard controller based endpoints.
3. Build standardized "TravelDeskTravelAgency" model.
4. Add new columns to travel agency creation and update forms and table.
5. Switch from using modals to using pages for travel agency create/update.
7. Remove now unused legacy "travel agents" routs from legacy "travel desk" router.
8. Remove some now unused legacy middlewares.
9. Add `validator.js` for front-end validation, and break current validators up into files. Back some of them with validator.js validators.
10. Remove some unused types from the User model.
11. Fix PlantUML rendering config, and add `dev plantuml-to-png <file-path>` helper in case VSCode plugin breaks again.

# Screenshots

http://localhost:8080/administration/travel-agencies
Travel Agencies admin table
![image](https://github.com/user-attachments/assets/5d6b518f-b327-472d-b79d-53edae2d3164)

http://localhost:8080/administration/travel-agencies/new
Travel Agency creation form
![image](https://github.com/user-attachments/assets/fc0280a8-898d-4b13-ad42-215eb5847b48)

http://localhost:8080/administration/travel-agencies/2/edit
Travel Agency edit form
![image](https://github.com/user-attachments/assets/ba76f4d3-7076-42fa-8da7-390d20ec35c3)

# Testing Instructions

1. Run the test suite via `dev test` (or `dev test_api`)
2. Boot the app via `dev up`
3. Log in to the app at http://localhost:8080
4. In the top right corner of the screen, click on the kebab menu, then select "Administration".
5. In the "Look up Management" section click on "travel agencies"
6. Add a new travel agency.
7. Check that validation works as expected.
8. Check that the agency renders correctly in the table.
9. Add a bunch more agencies.
10. Check that pagination works correctly on the table.
11. Check that edit redirects you to an edit page and that edit works correctly.
12. Note that "save" does not redirect you back to the table. This is a change from the previous modal behavior, but I found it less confusing, as the table might not display the entry you just edited.
13. Check that delete works as expected.
14. Note that delete will "restrict" if the agency is being used by a travel desk request.
15. Go to the Travel Desk via the top drop down nav.
16. Click to "edit" a travel desk request.
17. Check that, in the top right, you can assign a travel agency from the list of ones you just created.
18. Scroll down to the Upload PNR button and click on it.
19. Check that this also lets you edit the assigned travel agency.
